### PR TITLE
avoid unnecessary broadcasting in jax.numpy.vectorize

### PR DESCRIPTION
**Pull request by Gattocrucco**
Simple but very effective optimization of `jnp.vectorize`, which was marked `TODO`. Issue #14751.

### CODE REVIEW

File: jax/_src/numpy/vectorize.py [hunk 1]
```diff
@@ -278,36 +278,43 @@ def wrapped(*args):
...
-      vec_args.append(vec_arg)
-
-      vmap_count = len(vec_shape) - len(core_shape)
-      vmap_counts.append(vmap_count)
+      noncore_shape = arg.shape[:arg.ndim - len(core_dims)]
+
+      pad_ndim = len(broadcast_shape) - len(noncore_shape)
+      filled_shape = pad_ndim * (1,) + noncore_shape
+      assert len(filled_shape) == len(broadcast_shape)
+      assert jnp.broadcast_shapes(broadcast_shape, filled_shape) == broadcast_shape
```
Comment (jakevdp) [inline jax/_src/numpy/vectorize.py:296]: If these assertions are intended to catch a user error, we should instead raise an informative error in place of a raw assertion. If these checks are unrelated to user error, we could put them behind `if config.jax_enable_checks`, then the assertions would only be executed during testing, not at normal runtime.

File: jax/_src/numpy/vectorize.py [hunk 2]
```diff
@@ -278,36 +278,43 @@ def wrapped(*args):
...
+      else:
+        vectorized_func = api.vmap(vectorized_func, in_axes)
+    result = vectorized_func(*squeezed_args)
+
+    if not dims_to_expand:
+      return result
+    elif output_core_dims is not None and len(output_core_dims) > 1:
+      return tuple(jnp.expand_dims(r, axis=dims_to_expand) for r in result)
+    else:
+      return jnp.expand_dims(result, axis=dims_to_expand)
```
Comment (jakevdp) [inline jax/_src/numpy/vectorize.py:318]: Does this correctly handle the case where the function returns multiple values? e.g.
```python
@jnp.vectorize
def f(x):
  return x, x + 1
result = f(jnp.arange(4))
```

File: jax/_src/numpy/vectorize.py (hunk 1 above)
Comment (Gattocrucco) [reply jax/_src/numpy/vectorize.py:296]: They are internal checks, I add `if config.jax_enable_checks`.

File: jax/_src/numpy/vectorize.py (hunk 2 above)
Comment (Gattocrucco) [reply jax/_src/numpy/vectorize.py:318]: Yes, it works. This is tested in the unit tests. Multiple outputs are handled by the `elif output_core_dims ...` branch.

Comment (jakevdp) [reply jax/_src/numpy/vectorize.py:316]: My reading of the code is that if `signature` is not specified, `output_core_dims` will be `None` and so that branch will be skipped. What am I missing?

File: tests/lax_numpy_vectorize_test.py [hunk 3]
```diff
@@ -216,6 +217,13 @@ def test_inconsistent_output_size(self):
         ValueError, r"inconsistent size for core dimension 'n'"):
       f(jnp.zeros((2, 3)), jnp.zeros((3, 4)))
 
+  def test_expand_dims_multiple_outputs_no_signature(self):
+    f = jnp.vectorize(lambda x: (x, x))
+    x = jnp.arange(1)
+    xx = f(x)
+    self.assertAllClose(xx[0], x)
```
Comment (jakevdp) [inline tests/lax_numpy_vectorize_test.py:224]: Can we also do `self.assertIsInstance(xx, tuple)`?

Comment (Gattocrucco) [reply tests/lax_numpy_vectorize_test.py:224]: ok

**Review by jakevdp — CHANGES_REQUESTED** [review]
Looks good, thanks!

The last thing we need is to squash all changes into a single commit (see https://jax.readthedocs.io/en/latest/contributing.html#single-change-commits-and-pull-requests), and then we can get this change merged. Thanks!

**Review by jakevdp — APPROVED** [review]
